File: day17/cube.js
const { neighbourCoords } = require('./coord');

class Cube {
  constructor(value, coord) {
    this.coord = coord;
    this.value = value;
    this.nextValue = null;
    this.neighbourHoodCoords = neighbourCoords(coord, 1);
  }

  get isActive() {
    return this.value === '#';
  }

  setNextValue(grid3D) {
    const activeNeighbourCoords = this.neighbourHoodCoords.reduce((acc, nC) => grid3D.isActive(nC) ? acc + 1 : acc, 0);
    const status = this.isActive;
    if (status && (activeNeighbourCoords === 2 || activeNeighbourCoords === 3) ) {
      this.nextValue = '#';
    } else if (!status && activeNeighbourCoords === 3) {
      this.nextValue = '#';
    } else {
      this.nextValue = '.';
    }
  }

  update() {
    this.currentValue = this.nextValue;
    this.nextValue = null;
  }
}

module.exports = {
  Cube
}

File: day17-go/part2/cube.go
package main

type Cube struct {
	active          bool
	coord           Coord
	NeighbourCoords []Coord
}

func Default(coord Coord) *Cube {
	return New(coord, false)
}

func New(coord Coord, active bool) *Cube {
	neighbourCoords := coord.ComputeNeighbourCoords()

	return &Cube{
		active:          active,
		coord:           coord,
		NeighbourCoords: neighbourCoords,
	}
}

func (c *Cube) Hash() string {
	return c.coord.Hash()
}

func (c *Cube) Toggle() {
	c.active = !c.active
}

func (c *Cube) IsActive() bool {
	return c.active
}

func (c *Cube) Enable() {
	c.active = true
}

func (c *Cube) Disable() {
	c.active = false
}

func (c *Cube) String() string {
	if c.active {
		return "#"
	}
	return "."
}
